### ISSUE TITLE
Urlencoding of client identifier and secret when using HTTP basic auth in access_token!

### ISSUE THREAD

**Issue by krisvage** [nov/rack-oauth2#66]
The OAuth 2.0 standard, in section [2.3.1](https://tools.ietf.org/html/rfc6749#section-2.3.1) on client authentication using a client password, says the following:

> The client identifier is encoded using the
   "application/x-www-form-urlencoded" encoding algorithm per
   Appendix B, and the encoded value is used as the username; the client
   password is encoded using the same algorithm and used as the
   password.

We ran into this problem when integrating with an OpenID Connect provider which only allows fully qualified urls as client_id (eg. "https://www.example.com").
With the current code in [lib/rack/oauth2/client.rb](https://github.com/nov/rack-oauth2/blob/master/lib/rack/oauth2/client.rb#L85) we get this example:
```ruby
identifier = 'https://www.example.com'
secret = 'supersecret'
cred = ["#{identifier}:#{secret}"].pack('m').tr("\n", '')
# => "aHR0cHM6Ly93d3cuZXhhbXBsZS5jb206c3VwZXJzZWNyZXQ="
```
While from the standard one might expect:
```ruby
cred = ["#{CGI.escape(identifier)}:#{CGI.escape(secret)}"].pack('m').tr("\n", '')
# => "aHR0cHMlM0ElMkYlMkZ3d3cuZXhhbXBsZS5jb206c3VwZXJzZWNyZXQ="
```
[This comment in a Github issue for another project does a better job of explaining the ambiguity of the spec than I could.](https://github.com/GluuFederation/oxAuth/issues/677#issuecomment-348350250)

Of course, using a colon in the identifier, when rfc2617 explicitly forbids this, doesn't seem like the best idea from our provider. There are however other characters that would also be escaped by this, like the '!' used in the example in the issue above.

I would be happy to provide a pull request with the code for this, if you feel like this is a issue you would like to address. If this is deemed a breaking change, we could maybe include this as a separate explicit `:client_auth_method`?

**Comment by cedricpim** [comment]
> I would be happy to provide a pull request with the code for this, if you feel like this is a issue you would like to address. If this is deemed a breaking change, we could maybe include this as a separate explicit :client_auth_method?

Any chance this could be done? Or would such an MR be accepted (I could work on it). I am currently facing a problem where I am using this gem to connect to a third party and their server expect the old type of encoding, not the new one. I could monkeypatch it, but then that would prevent me from actually connecting correctly with third parties that implement this rfc.

**Comment by nov** [comment]
@cedricpim 
Does this works for you?
https://github.com/nov/rack-oauth2/pull/84

**Comment by cedricpim** [comment]
Yes, that would work. An accounting bookkeeping system: https://accounting2.twinfield.com/webservices/documentation/#/ApiReference/Authentication/OpenIdConnect.

**Comment by nov** [comment]
I've just released `v1.17.0` with this change.
thanks!
